tests/modules/mesoscope_splitting/test_tiff_splitter.py

```diff
@@ -0,0 +1,556 @@
+        z1 = roi_to_z_mapping[i_roi][1]
+        z1_values = np.linspace(z1-5.0, z1+5.0, 7)
+        z_values = []
+        for ii in range(len(z0_values)):
+            z_values.append([z0_values[ii], z1_values[ii]])
+
+        metadata = [{z_key: z_values},
+                    roi_metadata]
+        str_path = str(stack_path.resolve().absolute())
+        z_stack_path_to_metadata[str_path] = metadata
+
+        this_tiff = []
+        for i_z in range(len(z0_values)):
+            for i_repeat in range(n_repeats):
+                for ii in (0, 1):
+                    this_z = roi_to_z_mapping[i_roi][ii]
+                    if (i_roi, this_z) not in tiff_pages_lookup:
+                        tiff_pages_lookup[(i_roi, this_z)] = []
+                    page = rng.integers(0, 2**16-1, (24, 24)).astype(np.int16)
+                    page[i_roi:i_roi+5, i_roi:i_roi+5] = 0
+                    page[i_z:i_z+2, i_z:i_z+2] = 1000
+                    tiff_pages_lookup[(i_roi, this_z)].append(page)
+                    this_tiff.append(page)
+
+        tifffile.imsave(stack_path, this_tiff)
+
+    return (z_stack_path_to_metadata,
+            tiff_pages_lookup)
+
+
+@pytest.mark.parametrize(
+    "z_value_list, use_zs",
+    [(list(range(8)), True),
+     (list(range(8)), True),
+     (list(range(6)), True),
+     ((0, 2, 3, 5, 7, 4, 6, 11), True),
+     ((0, 2, 3, 5, 7, 4, 6, 11), True),
+     ((0, 2, 3, 5, 7, 4, 6, 5), True),
+     ((0, 2, 3, 5, 7, 4, 6, 5), True),
+     ((0, 4, 5, 2, 8, 7), True),
+     (list(range(8)), False),
+     (list(range(8)), False),
+     (list(range(6)), False),
+     ((0, 2, 3, 5, 7, 4, 6, 11), False),
+     ((0, 2, 3, 5, 7, 4, 6, 11), False),
+     ((0, 2, 3, 5, 7, 4, 6, 5), False),
+     ((0, 2, 3, 5, 7, 4, 6, 5), False),
+     ((0, 4, 5, 2, 8, 7), False)])
+def test_z_stack_splitter(tmp_path_factory,
+                          z_value_list,
+                          use_zs):
+    """
+    Test that _get_pages and write_output_file behave properly
+    for zstack_splitter
+    """
+    tmpdir = pathlib.Path(tmp_path_factory.mktemp('z_stack_test'))
+
+    n_z_per_roi = 2  # a requirement of the z-stacks
+    n_rois = len(z_value_list) // n_z_per_roi
+    roi_to_z_mapping = []
+    for i_roi in range(n_rois):
+        i0 = i_roi*n_z_per_roi
+        i1 = i0 + n_z_per_roi
+        these_zs = z_value_list[i0:i1]
+        roi_to_z_mapping.append(these_zs)
+
+    dataset = _create_z_stack_tiffs(
+                    tmpdir=tmpdir,
+                    roi_to_z_mapping=roi_to_z_mapping,
+                    use_zs=use_zs)
+
+    z_stack_path_to_metadata = dataset[0]
+    tiff_pages_lookup = dataset[1]
+
+    def mock_read_metadata(tiff_path):
+        str_path = str(tiff_path.resolve().absolute())
+        return z_stack_path_to_metadata[str_path]
+
+    to_replace = 'ophys_etl.modules.mesoscope_splitting.'
+    to_replace += 'tiff_metadata._read_metadata'
+    with patch(to_replace,
```

Comment:
This might be relevant to other patches as well but when do you stop needing this patch? My habit is to leave the with statement as soon as I don't need it.

Reply:
Done